src/main/java/com/_up/megastore/services/implementations/UserService.java

```diff
@@ -1,68 +1,79 @@
 import com._up.megastore.services.mappers.UserMapper;
+import java.time.LocalDateTime;
 import java.util.UUID;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.stereotype.Service;
 
 import java.util.NoSuchElementException;
 
+import static java.util.UUID.randomUUID;
+
 @Service
 public class UserService implements IUserService {
 
   private final IEmailService emailService;
   private final IUserRepository userRepository;
+  private final ITokenRepository tokenRepository;
   private final PasswordEncoder passwordEncoder;
 
   @Value("${frontend.url}")
   private String frontendURL;
 
   public UserService(IUserRepository userRepository, PasswordEncoder passwordEncoder,
-      IEmailService emailService) {
+      IEmailService emailService, ITokenRepository tokenRepository) {
     this.emailService = emailService;
     this.passwordEncoder = passwordEncoder;
     this.userRepository = userRepository;
+    this.tokenRepository = tokenRepository;
   }
 
   @Override
   public void saveUser(SignUpRequest createUserRequest) {
     ifEmailAlreadyExistsThrowException(createUserRequest.email());
     ifUsernameAlreadyExistsThrowException(createUserRequest.username());
     User newUser = createUser(createUserRequest);
-    sendActivationEmail(newUser);
+    UUID activationTokenUUID = generateActivationToken(newUser);
+    sendActivationEmail(newUser,activationTokenUUID);
     userRepository.save(newUser);
   }
 
   @Override
   public User findUserByUsernameOrThrowException(String username) {
     return userRepository.findByUsername(username)
-            .orElseThrow(() -> new NoSuchElementException("User with username "+username+" does not exist."));
+            .orElseThrow(() -> new NoSuchElementException("User with username " + username + " does not exist."));
   }
 
   @Override
   public void activateUser(UUID userId, UUID activationToken) {
     User user = findUserToActivateOrThrowException(userId, activationToken);
+    if(isTokenExpired(activationToken)){
+      throw new IllegalArgumentException("Token has expired");
+    }
     user.setActivated(true);
     userRepository.save(user);
     sendWelcomeEmail(user);
   }
 
-  User createUser(SignUpRequest createUserRequest) {
+  private User createUser(SignUpRequest createUserRequest) {
     User user = UserMapper.toUser(createUserRequest);
     String passwordEncoded = passwordEncoder.encode(createUserRequest.password());
     user.setPassword(passwordEncoded);
     return user;
   }
 
-  private void sendActivationEmail(User user) {
+  private void sendActivationEmail(User user, UUID activationToken) {
     String activationUrl = frontendURL + "/auth/activate?userId="
-        + user.getUserId() + "&activationToken=" + user.getActivationToken();
+        + user.getUserId() + "&activationToken=" + findActivationTokenByTokenUUIDOrThrowException(activationToken).getTokenId();
```

Comment:
Si lo acabamos de generar, por qué lo vamos a buscar a la BD?

```suggestion
        + user.getUserId() + "&activationToken=" + activationToken;
```